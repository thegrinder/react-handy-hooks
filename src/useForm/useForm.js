import { useState, useCallback, useMemo } from 'react';
import equal from 'fast-deep-equal';

const useForm = ({ initialValues, validate, onSubmit, submitting = false }) => {
  const [values, updateValues] = useState(initialValues);
  const [errors, updateErrors] = useState(
    validate ? validate(initialValues) : {}
  );
  const [touched, updateTouched] = useState({});

  const runValidation = useCallback(
    formValues => {
      if (validate) {
        updateErrors(validate(formValues));
      }
    },
    [validate]
  );

  const invalid = useMemo(() => !!Object.keys(errors).length, [errors]);

  const pristine = useMemo(() => equal(initialValues, values), [
    initialValues,
    values,
  ]);

  const getFieldProps = useCallback(
    name => ({
      submitting,
      error: errors[name],
      touched: !!touched[name],
      value: values[name],
      onChange(e) {
        const newValues = { ...values, [name]: e.target.value };
        updateValues(newValues);
        runValidation(newValues);
      },
      onBlur() {
        updateTouched({ ...touched, [name]: true });
      },
    }),
    [errors, runValidation, submitting, touched, values]
  );

  const handleSubmit = useCallback(
    e => {
      e.preventDefault();
      if (!invalid && onSubmit) {
        onSubmit(values);
      }
    },
    [invalid, onSubmit, values]
  );

  const resetForm = useCallback(() => {
    updateValues(initialValues);
    updateErrors({});
    updateTouched({});
  }, [initialValues]);

  return {
    getFieldProps,
    invalid,
    pristine,
    submitting,
    handleSubmit,
    resetForm,
  };
};

export default useForm;
