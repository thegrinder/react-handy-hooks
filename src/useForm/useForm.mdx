---
name: useForm
menu: hooks
---

# useForm

useForm is a simple react hook that allows you to build forms quickly.



## Arguments

useForm accepts an object with `initialValues`, `validate`, and `onSubmit` properties as its only argument.

### **initialValues**

It's the only required property. It's an object consisting of field names and its corresponding values.

### **validate**

It's a function that returns errors object where the keys are the field names and the values - the corresponding errors.

### **onSubmit**

It's a function called with form values.



## Returned object

useForm returns an object with the following properties, `getFieldProps`, `invalid`, `pristine`, `handleSubmit`and `resetForm`.

### **getFieldProps**

It's a function that takes in a field's name, and returns all the handlers and flags needed to to create a form field.
Here is the list of returned handlers/flags:

- error
- touched
- submitting
- value
- onChange
- onBlur

### **invalid**

It's a boolean that signals whether the form is invalid.

### **pristine**

It's a boolean that signals whether the form values are equal to its initial values.

### **handleSubmit**

It's a wrapper function around the given onSubmit that normally should be applied to form's onSubmit.

### **resetForm**

It's a function that resets the form to its initialState;.



## Example

```js
const Input = ({ error, touched, value, onChange, onBlur }) => (
  <div>
    <input
      onChange={onChange}
      onBlur={onBlur}
      value={value}
    />
    <div>
      {touched && error && <span style={{ color: 'red' }}>{error}</span>}
    </div>
  </div>
);

const validate = (values) => {
  let errors = {};
  if (values.name === '') {
    errors.name = 'required';
  }
  return errors;
};

const initialValues = {
  name: '',
};

const onSubmit = values => {
  alert(JSON.stringify(values));
};

const Example = () => {
  const { pristine, invalid, getFieldProps, handleSubmit } = useForm({
    initialValues,
    validate,
    onSubmit,
  });
  const { error, touched, value, onChange, onBlur } = getFieldProps('name');
  return (
    <form onSubmit={handleSubmit}>
      <Input
        error={error}
        touched={touched}
        value={value}
        onChange={onChange}
        onBlur={onBlur}
      />
      <div>
        <button disabled={pristine || invalid}>submit</button>
      </div>
    </form>
  );
};
```

import { Playground } from 'docz';
import { useForm } from '../';

<Playground>
  {() => {
    const Input = ({ error, touched, value, onChange, onBlur }) => (
      <div>
        <input
          onChange={onChange}
          onBlur={onBlur}
          value={value}
        />
        <div>
          {touched && error && <span style={{ color: 'red' }}>{error}</span>}
        </div>
      </div>
    );

    const validate = (values) => {
      let errors = {};
      if (values.name === '') {
        errors.name = 'required';
      }
      return errors;
    };

    const initialValues = {
      name: '',
    };

    const onSubmit = values => {
      alert(JSON.stringify(values));
    };

    const Example = () => {
      const { pristine, invalid, getFieldProps, handleSubmit } = useForm({
        initialValues,
        validate,
        onSubmit,
      });
      const { error, touched, value, onChange, onBlur } = getFieldProps('name');
      return (
        <form onSubmit={handleSubmit}>
          <Input
            error={error}
            touched={touched}
            value={value}
            onChange={onChange}
            onBlur={onBlur}
          />
          <div>
            <button disabled={pristine || invalid}>submit</button>
          </div>
        </form>
      );
    };

    return <Example />
  }}
</Playground>

