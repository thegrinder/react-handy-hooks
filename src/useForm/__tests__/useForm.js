import { renderHook, cleanup, act } from 'react-hooks-testing-library';
import useForm from '../useForm';

const initialValues = {
  name: 'name',
  lastName: 'lastName',
};
const onSubmit = jest.fn();
const validate = (values) => {
  const errors = {};
  if (values.name === '') {
    errors.name = 'required name';
  }
  if (values.lastName === '') {
    errors.lastName = 'required last name';
  }
  return errors;
};

describe('useForm', () => {
  afterEach(() => {
    cleanup();
    onSubmit.mockClear();
  });

  it('returns the correct object', () => {
    const { result } = renderHook(() => useForm({
      initialValues,
      onSubmit,
      validate,
    }));
    ['getFieldProps', 'invalid', 'pristine', 'handleSubmit', 'resetForm'].forEach((property) => {
      expect(result.current).toHaveProperty(property);
    });
    expect(result.current).toMatchSnapshot();
  });

  describe('invalid form flag', () => {
    it('should be false if no validate function is provided', () => {
      const { result } = renderHook(() => useForm({
        initialValues,
        onSubmit,
      }));
      expect(result.current.invalid).toEqual(false);
    });

    it('should be false if the fields are valid', () => {
      const { result } = renderHook(() => useForm({
        initialValues,
        onSubmit,
        validate,
      }));
      expect(result.current.invalid).toEqual(false);
    });

    it('should be true if at least one of fields is invalid', () => {
      const { result } = renderHook(() => useForm({
        initialValues: {
          ...initialValues,
          name: '',
        },
        onSubmit,
        validate,
      }));
      expect(result.current.invalid).toEqual(true);
    });
  });

  describe('pristine form flag', () => {
    it('should be true when the form is initialized', () => {
      const { result } = renderHook(() => useForm({
        initialValues,
        onSubmit,
        validate,
      }));
      expect(result.current.pristine).toEqual(true);
    });

    it('should be false if one of the field values has changed', () => {
      const { result } = renderHook(() => useForm({
        initialValues,
        onSubmit,
        validate,
      }));
      act(() => {
        result.current.getFieldProps('name').onChange({ target: { value: 'value' } });
      });
      expect(result.current.pristine).toEqual(false);
    });
  });

  describe('handleSubmit', () => {
    it('should fire provided onSubmit callback with form values', () => {
      const preventDefault = jest.fn();
      const { result } = renderHook(() => useForm({
        initialValues,
        onSubmit,
        validate,
      }));
      act(() => {
        result.current.handleSubmit({ preventDefault });
      });
      expect(preventDefault).toHaveBeenCalled();
      expect(onSubmit).toHaveBeenCalledWith(initialValues);
    });

    it('should not fire provided onSubmit callback if the form is invalid', () => {
      const preventDefault = jest.fn();
      const { result } = renderHook(() => useForm({
        initialValues: {
          ...initialValues,
          name: '',
        },
        onSubmit,
        validate,
      }));
      act(() => {
        result.current.handleSubmit({ preventDefault });
      });
      expect(onSubmit).not.toHaveBeenCalled();
    });
  });
});
