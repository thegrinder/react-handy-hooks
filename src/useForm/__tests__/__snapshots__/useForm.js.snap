// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useForm getFieldProps should return the correct object 1`] = `
Object {
  "error": undefined,
  "onBlur": [Function],
  "onChange": [Function],
  "submitting": false,
  "touched": false,
  "value": "name",
}
`;

exports[`useForm getFieldProps should return the correct object 2`] = `
Object {
  "error": undefined,
  "onBlur": [Function],
  "onChange": [Function],
  "submitting": false,
  "touched": false,
  "value": "lastName",
}
`;

exports[`useForm should return the correct object 1`] = `
Object {
  "getFieldProps": [Function],
  "handleSubmit": [Function],
  "invalid": false,
  "pristine": true,
  "resetForm": [Function],
}
`;
